### PR TITLE
Nested Conditional support and context cleanup

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
* Fixed an issue when starting or completing branches - no underscore needed for dataclasses.
* Fixed issue with mode comparison when deciding whether or not branch contents should run.
* Added a new state to the branchevalmode.  Previously, only tasks/lps/wfs could be in the `then` portion of a conditional.  Now conditional can be in the `then` part of the conditional.  Tasks/WFs should detect if they're in an untaken branch and not run... so do conditionals now.

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1052

Also dependent on PR - https://github.com/flyteorg/flytepropeller/pull/264

### CODE REVIEW

**Comment by wild-endeavor** [comment]
+1 by me.

File: flytekit/core/context_manager.py [hunk 1]
```diff
@@ -369,22 +378,23 @@ def build(self) -> FlyteContext:
             )
 
         def enter_conditional_section(self) -> "Builder":
-            if self.in_a_condition:
-                raise NotImplementedError("Nested branches are not yet supported!")
+            # if self.in_a_condition:
```
Comment (wild-endeavor) [inline flytekit/core/context_manager.py:381]: can nix this too.

Comment (kumare3) [reply flytekit/core/context_manager.py:381]: ohh ya

**Comment by wild-endeavor** [comment]
@kumare3 you have to +1 and review this...

**Review by kumare3 — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]
